src/db_reader.rs

```diff
@@ -0,0 +1,1083 @@
+                .clone()
+        } else {
+            let options = CheckpointOptions {
+                lifetime: Some(options.checkpoint_lifetime),
+                ..CheckpointOptions::default()
+            };
+            manifest.write_checkpoint(None, &options).await?
+        };
+        Ok(checkpoint)
+    }
+
+    async fn get_with_options<K: AsRef<[u8]> + Send>(
+        &self,
+        key: K,
+        options: &ReadOptions,
+    ) -> Result<Option<Bytes>, SlateDBError> {
+        self.reader.get_with_options(key, options).await
+    }
+
+    async fn scan_with_options(
+        &self,
+        range: BytesRange,
+        options: &ScanOptions,
+    ) -> Result<DbIterator, SlateDBError> {
+        self.reader.scan_with_options(range, options).await
+    }
+
+    fn should_reestablish_checkpoint(&self, latest: &CoreDbState) -> bool {
+        let read_guard = self.state.read();
+        let current_state = read_guard.state.core();
+        latest.last_compacted_wal_sst_id > current_state.last_compacted_wal_sst_id
+            || latest.next_wal_sst_id > current_state.next_wal_sst_id
+            || latest.l0_last_compacted != current_state.l0_last_compacted
+    }
+
+    async fn replace_checkpoint(
+        &self,
+        stored_manifest: &mut StoredManifest,
+    ) -> Result<Checkpoint, SlateDBError> {
+        let current_checkpoint_id = self.state.read().checkpoint.id;
+        let options = CheckpointOptions {
+            lifetime: Some(self.options.checkpoint_lifetime),
+            ..CheckpointOptions::default()
+        };
+        stored_manifest
+            .replace_checkpoint(current_checkpoint_id, &options)
+            .await
+    }
+
+    async fn reestablish_checkpoint(&self, checkpoint: Checkpoint) -> Result<(), SlateDBError> {
+        let new_checkpoint_state = self.rebuild_checkpoint_state(checkpoint).await?;
+        let mut write_guard = self.state.write();
+        *write_guard = new_checkpoint_state;
+        Ok(())
+    }
+
+    async fn maybe_replay_new_wals(&self) -> Result<(), SlateDBError> {
+        let last_seen_wal_id = self.table_store.last_seen_wal_id().await?;
+        let last_replayed_wal_id = self.state.read().last_wal_id;
+        if last_seen_wal_id > last_replayed_wal_id {
+            let mut updated_checkpoint = self.state.read().clone();
+            let mut updated_state = updated_checkpoint.state.as_ref().clone();
+
+            Self::replay_wal_into(
+                Arc::clone(&self.table_store),
+                &self.options,
+                updated_checkpoint.state.core(),
+                &mut updated_state.imm_memtable,
+                true,
+            )
+            .await?;
+            updated_checkpoint.state = Arc::new(updated_state);
+            let mut write_guard = self.state.write();
+            *write_guard = updated_checkpoint;
+        }
+        Ok(())
+    }
+
+    async fn build_initial_checkpoint_state(
```

Comment:
nit: can this and the fn below just call some common fn for building the state given the current `imm_memtable` and a checkpoint id?